### PR TITLE
Make CLI configuration for TTS Cloud and TTS Community Edition easier

### LINKED ISSUE #3591: Check and clean url in `ttn-lw-cli use`
#### Summary

It would be great to process the url passed to `ttn-lw-cli use` and trim the protocol and trailing slashes for example

#### Why do we need this?

`ttn-lw-cli use https://thethings.example.com` will produce requests to `https://https://thethings.example.com`
`ttn-lw-cli use thethings.example.com/` will produce `https://thethings.example.com//...`

I assume passing a port also doesn't work

#### What is already there? What do you see now?

`ttn-lw-cli use` doesn't care what url is fed to it

#### What is missing? What do you want to see?

Some url checking/cleaning

#### Can you do this yourself and submit a Pull Request?

@neoaggelos

### LINKED ISSUE #4016: Simplify CLI configuration for TTN v3 and TTSC
#### Summary

Simplify CLI configuration for The Things Network V3 and The Things Stack Cloud deployments.

#### Why do we need this?

All TTN clusters share the Identity Server from eu1. This breaks the `ttn-lw-cli use nam1.cloud.thethings.network` command, because it sets the Identity Server address to `nam1.cloud.thethings.network`, which is not correct.

#### What is already there? What do you see now?

`ttn-lw-cli use [cluster]` command can be used to configure the CLI for use with a single-host deployment at `[cluster]`

#### What is missing? What do you want to see?

Make it straightforward to use the `ttn-lw-cli use` command with the TTN and TTSC deployments. 

#### Environment

3.12

#### How do you propose to implement this?

The TTN and TTSC deployments use well-known host addresses, and we do not expect them to change frequently. Ideally, we could SRV records for this (https://github.com/TheThingsNetwork/lorawan-stack/issues/3964), but we can also start with parsing the specified host in the CLI and:

- If we match a `([a-z0-9]+).cloud.thethings.network`, then set Identity Server and OAuth address to `eu1.cloud.thethings.network`.
- If we match a `([a-z0-9\-_]+).([a-z0-9]+).cloud.thethings.industries`, then set Identity Server and OAuth address to `\1.eu1.cloud.thethings.industries`.

If the region does not match any of the known deployments, we could also print a warning message.

#### How do you propose to test this?

Testing locally should be enough

#### Can you do this yourself and submit a Pull Request?

Yes

### PR DESCRIPTION
**Pull request by neoaggelos**
#### Summary

Closes #4016
Closes #3591

#### Changes

- When passing a TTS Cloud or TTS Community Edition hostname, configure OAuth and Identity Server address.
- Validate hostname passed to `ttn-lw-cli use` command.

#### Testing

```
$ ttn-lw-cli use nam1.cloud.thethings.network
INFO    Configuring for The Things Stack Community Edition      {"cluster_id": "nam1"}
INFO    Set OAuth Server address        {"address": "https://eu1.cloud.thethings.network/oauth"}
INFO    Set Identity Server gRPC address        {"address": "eu1.cloud.thethings.industries:8884"}
INFO    Config file for nam1.cloud.thethings.network written in .ttn-lw-cli.yml

$ ttn-lw-cli use tti.nam1.cloud.thethings.industries
INFO    Configuring for The Things Stack Cloud  {"cluster_id": "nam1", "tenant_id": "tti"}
INFO    Set OAuth Server address        {"address": "https://tti.eu1.cloud.thethings.industries/oauth"}
INFO    Set Identity Server gRPC address        {"address": "tti.eu1.cloud.thethings.industries:8884"}
INFO    Config file for tti.nam1.cloud.thethings.industries written in .ttn-lw-cli.yml
```

#### Notes for Reviewers

Changed the implementation to the one suggested by the review comments.

#### Checklist

- [X] Scope: The referenced issue is addressed, there are no unrelated changes.
- [X] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [X] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [X] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: cmd/ttn-lw-cli/commands/use.go [hunk 1]
```diff
@@ -25,23 +25,169 @@ import (
...
+
+			cluster = strings.TrimSuffix(strings.TrimSpace(cluster), ".cloud.thethings.industries")
+			if cluster == "" {
+				logger.Warnf("Please specify the The Things Stack Cloud cluster you want to connect to. Known cluster names: %v", knownCloudClusters)
+				logger.Warn("For more information, see https://console.cloud.thethings.industries/")
+				cmd.Usage()
+				return errNoCluster.New()
+			} else if !sliceContains(knownCommunityClusters, cluster) {
+				logger.Warnf("Unknown cluster name '%s'! Your configuration may be incorrect!", cluster)
+				logger.Warnf("Unless you know what you are doing, please use one of the following cluster names: %v", knownCloudClusters)
```
Comment (htdvisser) [inline cmd/ttn-lw-cli/commands/use.go]: Maybe mention that the CLI may need to be updated?

File: cmd/ttn-lw-cli/commands/use.go [hunk 2]
```diff
@@ -25,23 +25,169 @@ import (
...
+				logger.Warn("For more information, see https://console.cloud.thethings.industries/")
+			}
+			if !identifierRegex.MatchString(tenant) {
+				logger.Warnf("Invalid tenant identifier '%s'! Your configuration is incorrect!", tenant)
+				logger.Warn("For more information, see https://console.cloud.thethings.industries/")
```
Comment (htdvisser) [inline cmd/ttn-lw-cli/commands/use.go]: That page doesn't document anything about tenant identifiers

**Review by htdvisser — COMMENTED** [review]
TBH I would prefer just updating the normal `use` command to recognize anything that ends with `cloud.thethings.network`/`cloud.thethings.industries` and make sure to do the right thing then.

**Review by adriansmares — COMMENTED** [review]
Nothing to add.

**Comment by neoaggelos** [comment]
> TBH I would prefer just updating the normal use command to recognize anything that ends with cloud.thethings.network/cloud.thethings.industries and make sure to do the right thing then.

As stated in the issue, I think it feels better to have separate commands here, as it clearly conveys the intention (use TTS Cloud, use TTS Community), therefore we can print messages that are far more helpful for new users. There are also a lot fewer edge-cases to solve.

This is intended for new-comers, who may even misspell cloud.thethings.network. Having a helpful warning message is better than silently accepting it and failing afterwards.

**Comment by htdvisser** [comment]
> As stated in the issue

There was zero discussion in that issue, so it's not like we decided on anything beforehand.

> it clearly conveys the intention (use TTS Cloud, use TTS Community)

Specifying the address also does that

> we can print messages that are far more helpful for new users

Like what?

> There are also a lot fewer edge-cases to solve.

Like what?

> This is intended for new-comers, who may even misspell cloud.thethings.network

newcomers who may misspell `cloud.thethings.network` should probably not use our CLI

**Comment by neoaggelos** [comment]
> There was zero discussion in that issue, so it's not like we decided on anything beforehand.

Indeed, which is why I implemented my initial suggestion and asked for feedback if it actually improves the experience for new users.

> > it clearly conveys the intention (use TTS Cloud, use TTS Community)
> 
> Specifying the address also does that

Not if users do not know if the address. For us, `cloud.thethings.network` and `cloud.thethings.industries` is well-known and straightforward, but people may be confused and specify something like `console.thethingsnetwork.org` instead.

> > we can print messages that are far more helpful for new users
> 
> Like what?

Like a warning/error if the community cluster specified does not exist.

> 
> > There are also a lot fewer edge-cases to solve.
> 
> Like what?

Like trying to override the gRPC port / oauth server address when configuring for TTS Cloud/CE.

> 
> > This is intended for new-comers, who may even misspell cloud.thethings.network
> 
> newcomers who may misspell `cloud.thethings.network` should probably not use our CLI

Newcomers who may misspell `cloud.thethings.network` might do it because of confusion, and not inability to type a domain name.

All in all, this is to help users from the community that are trying to get started with using the nam1/au1 clusters and cannot get the configuration right. If we decide that it is better to add more logic to the existing `ttn-lw-cli use` command and keep pointing users to the documentation if they do not get it right, then this is fine by me as well.

**Comment by johanstokking** [comment]
The barrier should be lowered by helping people entering the right stuff in their terminal.

You are trying to lower the barrier by reducing characters to enter and adding client side validation (like clusters). But that is not how CLI works and should work. We have a Console for people who don't dare to use CLI. For those who do want to use the CLI, we need proper documentation and consistency, not magic.

**Comment by htdvisser** [comment]
This did not make it for the v3.13.3 deadline, so pushing to v3.14.0

**Review by htdvisser — CHANGES_REQUESTED** [review]
Took another look, but haven't changed my mind. This functionality belongs in the existing `use` command. Please make that command recognize:

- `<tenant>.<cluster>.cloud.thethings.industries`
- `<cluster>.cloud.thethings.industries` (error: missing tenant)
- `<cluster>.cloud.thethings.network`

And please use the `cluster` regex `(nam|sam|eu|af|as|au)[0-9]` instead of hardcoding clusters.

File: cmd/ttn-lw-cli/commands/use.go [hunk 3]
```diff
@@ -25,23 +25,40 @@ import (
...
+	validHostnameRegex             = regexp.MustCompile(`^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])$`)
+	communityClusterRegex          = regexp.MustCompile(`^.cloud.thethings.network$`)
+	cloudClusterRegex              = regexp.MustCompile(`^([a-z][a-z0-9-]{2,}).([a-z][a-z0-9-]{2,}).cloud.thethings.industries$`)
+	cloudClusterMissingTenantRegex = regexp.MustCompile(`^([a-z][a-z0-9-]{2,}).cloud.thethings.industries$`)
+
+	useCommand = &cobra.Command{
+		Use:               "use",
 		Aliases:           []string{"generate-configuration", "generate-cfg"},
-		Short:             "Use",
+		Example:           `use localhost --insecure --oauth-server-address http://localhost:1885/oauth`,
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/use.go]: This is a bad example

File: cmd/ttn-lw-cli/commands/use.go [hunk 4]
```diff
@@ -25,23 +25,40 @@ import (
...
+
 var (
 	errNoHost     = errors.DefineInvalidArgument("no_host", "no host set")
 	errFileExists = errors.DefineAlreadyExists("file_exists", "`{file}` exists")
 	errFailWrite  = errors.DefinePermissionDenied("fail_write", "failed to write `{file}`")
-	useCommand    = &cobra.Command{
-		Use:               "use [host]",
+
+	validHostnameRegex             = regexp.MustCompile(`^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])$`)
+	communityClusterRegex          = regexp.MustCompile(`^.cloud.thethings.network$`)
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/use.go]: This doesn't match `eu1.cloud.thethings.network`

File: cmd/ttn-lw-cli/commands/use.go [hunk 5]
```diff
@@ -25,23 +25,40 @@ import (
...
+	cloudClusterRegex              = regexp.MustCompile(`^([a-z][a-z0-9-]{2,}).([a-z][a-z0-9-]{2,}).cloud.thethings.industries$`)
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/use.go]: This should be the tenant regex + region (`nam|sam|eu|af|as|au`) with number + suffix

File: cmd/ttn-lw-cli/commands/use.go [hunk 6]
```diff
@@ -25,23 +25,40 @@ import (
...
+	cloudClusterMissingTenantRegex = regexp.MustCompile(`^([a-z][a-z0-9-]{2,}).cloud.thethings.industries$`)
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/use.go]: Or make the tenant ID in the previous optional and check if that group exists or not

File: cmd/ttn-lw-cli/commands/use.go [hunk 7]
```diff
@@ -25,23 +25,40 @@ import (
...
+}
+
 var (
 	errNoHost     = errors.DefineInvalidArgument("no_host", "no host set")
 	errFileExists = errors.DefineAlreadyExists("file_exists", "`{file}` exists")
 	errFailWrite  = errors.DefinePermissionDenied("fail_write", "failed to write `{file}`")
-	useCommand    = &cobra.Command{
-		Use:               "use [host]",
+
+	validHostnameRegex             = regexp.MustCompile(`^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])$`)
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/use.go:47]: Is this https://datatracker.ietf.org/doc/html/rfc1123? Please add reference here.

File: cmd/ttn-lw-cli/commands/use.go [hunk 8]
```diff
@@ -53,6 +70,11 @@ var (
 			overwrite, _ := cmd.Flags().GetBool("overwrite")
 
 			host := args[0]
+			if !validHostnameRegex.MatchString(host) {
+				logger.Warn("Specify the hostname only, for example `eu1.cloud.thethings.network`")
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/use.go]: I don't think we use backticks in log messages

File: cmd/ttn-lw-cli/commands/utils.go [hunk 9]
```diff
@@ -111,3 +111,27 @@ func parsePayloadFormatterParameterFlags(prefix string, formatters *ttnpb.Messag
 	}
 	return util.NormalizePaths(paths), nil
 }
+
+// usageTemplateWithoutGlobalFlags is the default usage message from Cobra, with
+// the Global Flags section stripped away.
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/utils.go]: Why?

Comment (neoaggelos) [reply cmd/ttn-lw-cli/commands/utils.go]: Global flags are ignored in the `use` command, and may be misleading for the users.

We can just revert if this does not help.

Comment (htdvisser) [reply cmd/ttn-lw-cli/commands/utils.go]: I agree that some of those global flags can be confusing, but then I'd prefer fixing that for all commands in a separate PR.

**Review by adriansmares — APPROVED** [review]

File: cmd/ttn-lw-cli/commands/use.go (hunk 6 above)
Comment (johanstokking) [reply cmd/ttn-lw-cli/commands/use.go]: This is missing region right?

File: cmd/ttn-lw-cli/commands/use.go (hunk 8 above)
Comment (johanstokking) [reply cmd/ttn-lw-cli/commands/use.go]: This is not addressed, please change or comment here

Comment (htdvisser) [reply cmd/ttn-lw-cli/commands/use.go]: You don't need to log a warning if you also return an error.

File: cmd/ttn-lw-cli/commands/use.go [hunk 10]
```diff
@@ -53,6 +70,11 @@ var (
...
+				return fmt.Errorf("host %q is not a valid hostname", host)
```
Comment (htdvisser) [inline cmd/ttn-lw-cli/commands/use.go]: Better even would be to define a TTN error

File: cmd/ttn-lw-cli/commands/use.go [hunk 11]
```diff
@@ -92,6 +114,28 @@ var (
 				return fileName, nil
 			}
 
+			if cloudClusterMissingTenantRegex.MatchString(host) {
+				logger.Warnf("Specify the tenant ID for The Things Stack Cloud, for example %q", "my-tenant.eu1.cloud.thethings.industries")
+				return fmt.Errorf("missing tenant ID")
```
Comment (htdvisser) [inline cmd/ttn-lw-cli/commands/use.go]: Better even would be to define a TTN error

File: cmd/ttn-lw-cli/commands/use.go [hunk 12]
```diff
@@ -25,23 +25,42 @@ import (
...
 	"go.thethings.network/lorawan-stack/v3/pkg/rpcmiddleware/discover"
 	"gopkg.in/yaml.v2"
 )
 
 const configFileName = ".ttn-lw-cli.yml"
 
+func useCommandWriteConfigFlags() *pflag.FlagSet {
+	flagSet := &pflag.FlagSet{}
+	flagSet.Bool("user", false, "Write config file in user config directory")
+	flagSet.Bool("overwrite", false, "Overwrite existing config files")
```
Comment (htdvisser) [inline cmd/ttn-lw-cli/commands/use.go]: I suppose you can move these flags back to the bottom now that we're back to a single command

**Review by htdvisser — COMMENTED** [review]
This diff is getting smaller and smaller. Great 😄

**Review by htdvisser — APPROVED** [review]

**Pull request merged**